b_cfn_opensearch_index/source/main/action.py

```diff
@@ -0,0 +1,216 @@
+import json
+import logging
+import re
+from secrets import token_urlsafe
+from typing import Any, Dict, NamedTuple, Optional
+
+import boto3
+
+from b_cfn_opensearch_index_layer.opensearch_client import OpensearchClient
+
+logger = logging.getLogger()
+
+
+class ActionResult(NamedTuple):
+    """
+    Custom data to return back to CloudFormation service, physical resource id (can be empty).
+    """
+
+    data: Optional[Dict[Any, Any]]
+    resource_id: str = None
+
+
+class Actions:
+    def __init__(self, invocation_event: Dict[str, Any]) -> None:
+        self.__properties: Dict[str, Any] = invocation_event['ResourceProperties']
+        self.__old_properties: Optional[Dict[str, Any]] = invocation_event.get('OldResourceProperties')
+        self.__resource_id: Optional[str] = invocation_event.get('PhysicalResourceId')
+
+        self.index_prefix = self.__properties['IndexPrefix']
+        # Check does index prefix comply with Opensearch index naming rules.
+        self.validate_index_prefix(index_prefix=self.index_prefix)
+
+        self.opensearch_domain_endpoint = self.__properties['OpensearchDomainEndpoint']
+        self.index_settings = self.__properties.get('IndexSettings') or {}
+        self.index_mapping_settings = {
+            'properties': self.__properties.get('IndexMappingSettings') or {}
+        }
+
+        # Setup Opensearch client.
+        self.opensearch_client = OpensearchClient(
+            boto3_session=boto3.Session(),
+            opensearch_endpoint=self.opensearch_domain_endpoint
+        )
+
+    def create(self) -> ActionResult:
+        """
+        Creates an empty index applying index settings and mappings from invocation event Resource Properties.
+
+        :return: Create action result.
+        """
+
+        logger.info(f'Initiating resource creation with these properties: {json.dumps(self.__properties)}.')
+
+        # Create index name by concatinating index prefix and url safe random string.
+        index_name = f'{self.index_prefix}-{token_urlsafe(8)}'.lower()
+
+        self.__create(
+            index_name=index_name,
+            body={
+                'settings': self.index_settings,
+                'mappings': self.index_mapping_settings
+            }
+        )
+
+        return ActionResult(data={'IndexName': index_name}, resource_id=f'{self.opensearch_domain_endpoint}||{index_name}')
```

Comment:
Move this to separate class: f'{self.opensearch_domain_endpoint}||{index_name}' - e.g. class OpensearchIndexId which is going to either build the id from endpoint and name or resolve endpoint and name from given id.

Reply:
And docstrings and comments are mandatory.